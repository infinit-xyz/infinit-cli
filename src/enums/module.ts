export enum PROTOCOL_MODULE {
  // init_capital = 'init_capital',
  // aave_v2 = 'aave_v2',
  token = 'token',
  aave_v3 = 'aave-v3',
  uniswap_v3 = 'uniswap-v3',
}
