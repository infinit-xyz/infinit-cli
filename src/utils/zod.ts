import { pipeInto } from 'ts-functional-pipe'
import { z } from 'zod'

export function zodGetDefaults<Schema extends z.AnyZodObject>(schema: Schema) {
  const entriesHandler = (entries: [string, z.AnyZodObject][]) => {
    return entries.map(([key, shape]) => {
      if (shape.isOptional()) return [undefined, undefined] as [string | undefined, undefined]
      return [key, undefined] as [string | undefined, undefined]
    })
  }

  const entriesFilter = (entries: [string | undefined, undefined][]): [string, undefined][] => {
    return entries.filter(([key]) => key !== undefined) as [string, undefined][]
  }

  return pipeInto(schema.shape, Object.entries, entriesHandler, entriesFilter, Object.fromEntries)
}
