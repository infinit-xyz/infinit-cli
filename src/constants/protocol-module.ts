import { actions as aaveV3Actions } from '@infinit-xyz/aave-v3/actions'
import { actions as tokenActions } from '@infinit-xyz/token/actions'
import { actions as uniswapV3Actions } from '@infinit-xyz/uniswap-v3/actions'

import { PROTOCOL_MODULE } from '@enums/module'
import type { ModuleAction } from './protocol-module.type'

export const protocolModules = {
  [PROTOCOL_MODULE.token]: {
    key: PROTOCOL_MODULE.token,
    name: 'Token',
    description: 'Token',
    actions: tokenActions,
    libPath: '@infinit-xyz/token',
  },
  [PROTOCOL_MODULE.aave_v3]: {
    key: PROTOCOL_MODULE.aave_v3,
    name: 'Aave V3',
    description: 'Aave V3',
    actions: aaveV3Actions,
    libPath: '@infinit-xyz/aave-v3',
  },
  [PROTOCOL_MODULE.uniswap_v3]: {
    key: PROTOCOL_MODULE.uniswap_v3,
    name: 'Uniswap V3',
    description: 'Uniswap V3',
    actions: uniswapV3Actions,
    libPath: '@infinit-xyz/uniswap-v3',
  },
} satisfies Record<PROTOCOL_MODULE, ModuleAction>
