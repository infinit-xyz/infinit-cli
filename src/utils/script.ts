import type { InfinitAction } from '@infinit-xyz/core'
import { pipeInto } from 'ts-functional-pipe'

import { stringifyWithUndefined } from '@utils/json'
import { zodGetDefaults } from '@utils/zod'

export const generateScriptText = (infinitAction: InfinitAction, libPath: string, actionKey: string, deployerId?: string) => {
  const signers = infinitAction.signers.reduce<Record<string, string>>((acc, signer) => {
    acc[signer] = signer === 'deployer' ? (deployerId ?? '') : ''
    return acc
  }, {})

  const generatedParamsText = pipeInto(infinitAction.paramsSchema, zodGetDefaults, stringifyWithUndefined)
  const splittedParamsText = generatedParamsText.split('\n')
  const newSplistedParamsText = []

  for (const line of splittedParamsText) {
    const match = line.match(/"([^"]+)":/) // to get the key from the line in double quotes
    if (match) {
      const key = match[1]
      newSplistedParamsText.push('\n')
      newSplistedParamsText.push(`  // TODO: ${infinitAction.paramsSchema.shape[key].description ?? ''}`)
    }
    newSplistedParamsText.push(line)
  }

  const scriptText = `
import { ${infinitAction.actionClassName}, type actions } from '${libPath}/actions'
import type { z } from 'zod'

type Param = z.infer<typeof actions['${actionKey}']['paramsSchema']>

// TODO: Replace with actual params
const params: Param = ${newSplistedParamsText.join('\n')}

// TODO: Replace with actual signer id
const signer = ${JSON.stringify(signers, undefined, 2)}

export default { params, signer, Action: ${infinitAction.actionClassName} }

`

  return scriptText
}
