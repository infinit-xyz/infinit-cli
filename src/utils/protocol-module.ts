import { protocolModules } from '@constants/protocol-module'
import type { PROTOCOL_MODULE } from '@enums/module'

export const getProtocolModule = (input?: string): PROTOCOL_MODULE | undefined => {
  if (!input) return undefined

  let protocolKey: PROTOCOL_MODULE | undefined = undefined

  protocolKey = protocolModules[input as PROTOCOL_MODULE]?.key

  if (!protocolKey) {
    const protocolByKeyReplace = protocolModules[input.replaceAll('-', '_') as PROTOCOL_MODULE]?.key
    protocolKey = protocolByKeyReplace
  }

  if (!protocolKey) {
    const protocolByName = Object.values(protocolModules).find((protocol) => protocol.name.toLowerCase() === input.toLowerCase())?.key
    protocolKey = protocolByName
  }

  return protocolKey
}
