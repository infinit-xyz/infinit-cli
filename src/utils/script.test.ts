import { protocolModules } from '@constants/protocol-module'
import { PROTOCOL_MODULE } from '@enums/module'
import { generateScriptText } from '@utils/script'
import { zodGetDefaults } from '@utils/zod'
import { describe, expect, test } from 'vitest'
import type { AnyZodObject } from 'zod'

describe('generateScriptText', () => {
  const initialAction = protocolModules[PROTOCOL_MODULE.aave_v3].actions.init
  const libPath = protocolModules[PROTOCOL_MODULE.aave_v3].libPath
  const paramsSchema = initialAction.paramsSchema as AnyZodObject
  const generatedText = generateScriptText(initialAction, libPath, 'init')

  test('should generate script text correctly (partial match)', () => {
    expect(generatedText).toContain(`import { ${initialAction.actionClassName}, type actions } from '${libPath}/actions'`)

    const zodObjectDefault = zodGetDefaults(initialAction.paramsSchema)
    const allZodKeys = Object.keys(zodObjectDefault)

    expect(generatedText).toContain(`type Param = z.infer<typeof actions['init']['paramsSchema']>`)
    expect(generatedText).toContain(`const params: Param = {`)
    allZodKeys.forEach((key) => {
      expect(generatedText).toContain(`"${key}": undefined`)
      expect(generatedText).toContain(`// TODO: ${paramsSchema.shape[key].description ?? ''}`)
    })

    expect(generatedText).toContain(`const signer = {`)
    initialAction.signers.forEach((signer) => {
      expect(generatedText).toContain(`"${signer}": ""`)
    })

    expect(generatedText).toContain(`export default { params, signer, Action: ${initialAction.actionClassName} }`)
  })

  test('should generate script text correctly (exact match)', () => {
    expect(generatedText).toEqual(`
import { DeployAaveV3Action, type actions } from '@infinit-xyz/aave-v3/actions'
import type { z } from 'zod'

type Param = z.infer<typeof actions['init']['paramsSchema']>

// TODO: Replace with actual params
const params: Param = {


  // TODO: Unique identifier for the AAVE market e.g. 'INFINIT-LENDING'
  "marketId": undefined,


  // TODO: Unique identifier for the addresses provider e.g. 1n
  "providerId": undefined,


  // TODO: Address of the treasury owner, managing protocol funds e.g. '0x123...abc'
  "treasuryOwner": undefined,


  // TODO: Address of the owner managing the address provider e.g. '0x123...abc'
  "addressesProviderOwner": undefined,


  // TODO: Address of the owner managing the address provider registry e.g. '0x123...abc'
  "addressesProviderRegistryOwner": undefined,


  // TODO: Address of the owner of the wrapped token gateway e.g. '0x123...abc'
  "wrappedTokenGatewayOwner": undefined,


  // TODO: Address of the Access Control List admin, managing protocol access e.g. '0x123...abc'
  "aclAdmin": undefined,


  // TODO: Address of the funds admin, managing funds distribution e.g. '0x123...abc'
  "fundsAdmin": undefined,


  // TODO: Address of the pool admin, managing lending pool settings e.g. '0x123...abc'
  "poolAdmin": undefined,


  // TODO: Address of the emergency admin, handling emergencies e.g. '0x123...abc'
  "emergencyAdmin": undefined,


  // TODO: Total flash loan premium rate in bps e.g. 50n
  "flashloanPremiumsTotal": undefined,


  // TODO: Portion of flash loan premium for the protocol in bps e.g. 50n
  "flashloanPremiumsProtocol": undefined,


  // TODO: Address of the Chainlink aggregator proxy for price data e.g. '0x123...abc'
  "chainlinkAggProxy": undefined,


  // TODO: Address of the Chainlink ETH/USD price feed proxy e.g. '0x123...abc'
  "chainlinkETHUSDAggProxy": undefined,


  // TODO: List of supported asset addresses
  "assets": undefined,


  // TODO: List of price feed source addresses
  "sources": undefined,


  // TODO: Address of the fallback oracle for backup price feeds e.g. '0x123...abc'
  "fallbackOracle": undefined,


  // TODO: Base token address of the protocol e.g. '0x123...abc'
  "baseCurrency": undefined,


  // TODO: Smallest unit of the base currency e.g. 1n
  "baseCurrencyUnit": undefined,


  // TODO: Address of the wrapped native token e.g. '0x123...abc'
  "wrappedNativeToken": undefined
}

// TODO: Replace with actual signer id
const signer = {
  "deployer": ""
}

export default { params, signer, Action: DeployAaveV3Action }

`)
  })

  test('should generate script text correctly with deployer (exact match)', () => {
    const generatedTextWithDeployer = generateScriptText(initialAction, libPath, 'init', 'test-deployer')
    expect(generatedTextWithDeployer).toEqual(`
import { DeployAaveV3Action, type actions } from '@infinit-xyz/aave-v3/actions'
import type { z } from 'zod'

type Param = z.infer<typeof actions['init']['paramsSchema']>

// TODO: Replace with actual params
const params: Param = {


  // TODO: Unique identifier for the AAVE market e.g. 'INFINIT-LENDING'
  "marketId": undefined,


  // TODO: Unique identifier for the addresses provider e.g. 1n
  "providerId": undefined,


  // TODO: Address of the treasury owner, managing protocol funds e.g. '0x123...abc'
  "treasuryOwner": undefined,


  // TODO: Address of the owner managing the address provider e.g. '0x123...abc'
  "addressesProviderOwner": undefined,


  // TODO: Address of the owner managing the address provider registry e.g. '0x123...abc'
  "addressesProviderRegistryOwner": undefined,


  // TODO: Address of the owner of the wrapped token gateway e.g. '0x123...abc'
  "wrappedTokenGatewayOwner": undefined,


  // TODO: Address of the Access Control List admin, managing protocol access e.g. '0x123...abc'
  "aclAdmin": undefined,


  // TODO: Address of the funds admin, managing funds distribution e.g. '0x123...abc'
  "fundsAdmin": undefined,


  // TODO: Address of the pool admin, managing lending pool settings e.g. '0x123...abc'
  "poolAdmin": undefined,


  // TODO: Address of the emergency admin, handling emergencies e.g. '0x123...abc'
  "emergencyAdmin": undefined,


  // TODO: Total flash loan premium rate in bps e.g. 50n
  "flashloanPremiumsTotal": undefined,


  // TODO: Portion of flash loan premium for the protocol in bps e.g. 50n
  "flashloanPremiumsProtocol": undefined,


  // TODO: Address of the Chainlink aggregator proxy for price data e.g. '0x123...abc'
  "chainlinkAggProxy": undefined,


  // TODO: Address of the Chainlink ETH/USD price feed proxy e.g. '0x123...abc'
  "chainlinkETHUSDAggProxy": undefined,


  // TODO: List of supported asset addresses
  "assets": undefined,


  // TODO: List of price feed source addresses
  "sources": undefined,


  // TODO: Address of the fallback oracle for backup price feeds e.g. '0x123...abc'
  "fallbackOracle": undefined,


  // TODO: Base token address of the protocol e.g. '0x123...abc'
  "baseCurrency": undefined,


  // TODO: Smallest unit of the base currency e.g. 1n
  "baseCurrencyUnit": undefined,


  // TODO: Address of the wrapped native token e.g. '0x123...abc'
  "wrappedNativeToken": undefined
}

// TODO: Replace with actual signer id
const signer = {
  "deployer": "test-deployer"
}

export default { params, signer, Action: DeployAaveV3Action }

`)
  })
})
