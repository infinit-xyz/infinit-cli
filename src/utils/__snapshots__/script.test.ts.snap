// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateScriptText > should generate script text correctly (exact match) 1`] = `
"
import { DeployAaveV3Action, type actions } from '@infinit-xyz/aave-v3/actions'
import type { z } from 'zod'

type Param = z.infer<typeof actions['init']['paramsSchema']>

// TODO: Replace with actual params
const params: Param = {


  // TODO: Unique identifier for the AAVE market e.g. 'INFINIT-LENDING'
  "marketId": undefined,


  // TODO: Unique identifier for the addresses provider e.g. 1n
  "providerId": undefined,


  // TODO: Address of the treasury owner, managing protocol funds e.g. '0x123...abc'
  "treasuryOwner": undefined,


  // TODO: Address of the owner managing the address provider e.g. '0x123...abc'
  "addressesProviderOwner": undefined,


  // TODO: Address of the owner managing the address provider registry e.g. '0x123...abc'
  "addressesProviderRegistryOwner": undefined,


  // TODO: Address of the owner of the wrapped token gateway e.g. '0x123...abc'
  "wrappedTokenGatewayOwner": undefined,


  // TODO: Address of the owner managing the emission manager e.g. '0x123...abc'
  "emissionManagerOwner": undefined,


  // TODO: Address of the Access Control List admin, managing protocol access e.g. '0x123...abc'
  "aclAdmin": undefined,


  // TODO: Address of the funds admin, managing funds distribution e.g. '0x123...abc'
  "fundsAdmin": undefined,


  // TODO: Address of the pool admin, managing lending pool settings e.g. '0x123...abc'
  "poolAdmin": undefined,


  // TODO: Address of the emergency admin, handling emergencies e.g. '0x123...abc'
  "emergencyAdmin": undefined,


  // TODO: Address of the rewards admin, managing rewards distribution e.g. '0x123...abc'
  "rewardsAdmin": undefined,


  // TODO: Address of the rewards holder, e.g. '0x123...abc'
  "rewardsHolder": undefined,


  // TODO: Total flash loan premium rate in bps e.g. 50n
  "flashloanPremiumsTotal": undefined,


  // TODO: Portion of flash loan premium for the protocol in bps e.g. 50n
  "flashloanPremiumsProtocol": undefined,


  // TODO: Address of the Chainlink aggregator proxy for price data e.g. '0x123...abc'
  "chainlinkAggProxy": undefined,


  // TODO: Address of the Chainlink ETH/USD price feed proxy e.g. '0x123...abc'
  "chainlinkETHUSDAggProxy": undefined,


  // TODO: List of supported asset addresses
  "assets": undefined,


  // TODO: List of price feed source addresses
  "sources": undefined,


  // TODO: Base token address of the protocol e.g. '0x123...abc'
  "baseCurrency": undefined,


  // TODO: Smallest unit of the base currency e.g. 1n
  "baseCurrencyUnit": undefined,


  // TODO: Address of the wrapped native token e.g. '0x123...abc'
  "wrappedNativeToken": undefined
}

// TODO: Replace with actual signer id
const signer = {
  "deployer": ""
}

export default { params, signer, Action: DeployAaveV3Action }

"
`;

exports[`generateScriptText > should generate script text correctly with deployer (exact match) 1`] = `
"
import { DeployAaveV3Action, type actions } from '@infinit-xyz/aave-v3/actions'
import type { z } from 'zod'

type Param = z.infer<typeof actions['init']['paramsSchema']>

// TODO: Replace with actual params
const params: Param = {


  // TODO: Unique identifier for the AAVE market e.g. 'INFINIT-LENDING'
  "marketId": undefined,


  // TODO: Unique identifier for the addresses provider e.g. 1n
  "providerId": undefined,


  // TODO: Address of the treasury owner, managing protocol funds e.g. '0x123...abc'
  "treasuryOwner": undefined,


  // TODO: Address of the owner managing the address provider e.g. '0x123...abc'
  "addressesProviderOwner": undefined,


  // TODO: Address of the owner managing the address provider registry e.g. '0x123...abc'
  "addressesProviderRegistryOwner": undefined,


  // TODO: Address of the owner of the wrapped token gateway e.g. '0x123...abc'
  "wrappedTokenGatewayOwner": undefined,


  // TODO: Address of the owner managing the emission manager e.g. '0x123...abc'
  "emissionManagerOwner": undefined,


  // TODO: Address of the Access Control List admin, managing protocol access e.g. '0x123...abc'
  "aclAdmin": undefined,


  // TODO: Address of the funds admin, managing funds distribution e.g. '0x123...abc'
  "fundsAdmin": undefined,


  // TODO: Address of the pool admin, managing lending pool settings e.g. '0x123...abc'
  "poolAdmin": undefined,


  // TODO: Address of the emergency admin, handling emergencies e.g. '0x123...abc'
  "emergencyAdmin": undefined,


  // TODO: Address of the rewards admin, managing rewards distribution e.g. '0x123...abc'
  "rewardsAdmin": undefined,


  // TODO: Address of the rewards holder, e.g. '0x123...abc'
  "rewardsHolder": undefined,


  // TODO: Total flash loan premium rate in bps e.g. 50n
  "flashloanPremiumsTotal": undefined,


  // TODO: Portion of flash loan premium for the protocol in bps e.g. 50n
  "flashloanPremiumsProtocol": undefined,


  // TODO: Address of the Chainlink aggregator proxy for price data e.g. '0x123...abc'
  "chainlinkAggProxy": undefined,


  // TODO: Address of the Chainlink ETH/USD price feed proxy e.g. '0x123...abc'
  "chainlinkETHUSDAggProxy": undefined,


  // TODO: List of supported asset addresses
  "assets": undefined,


  // TODO: List of price feed source addresses
  "sources": undefined,


  // TODO: Base token address of the protocol e.g. '0x123...abc'
  "baseCurrency": undefined,


  // TODO: Smallest unit of the base currency e.g. 1n
  "baseCurrencyUnit": undefined,


  // TODO: Address of the wrapped native token e.g. '0x123...abc'
  "wrappedNativeToken": undefined
}

// TODO: Replace with actual signer id
const signer = {
  "deployer": "test-deployer"
}

export default { params, signer, Action: DeployAaveV3Action }

"
`;
